pilot/pkg/model/context.go

```diff
@@ -146,6 +151,77 @@ func (e *Environment) Version() string {
 	return ""
 }
 
+func (e *Environment) InitClusterLocalHosts() []host.Name {
+	// Create the default list of cluster-local hosts.
+	domainSuffix := e.GetDomainSuffix()
+	defaultClusterLocalHosts := make([]host.Name, 0)
+	for _, n := range defaultClusterLocalNamespaces {
+		defaultClusterLocalHosts = append(defaultClusterLocalHosts, host.Name("*."+n+".svc."+domainSuffix))
+	}
+	for _, s := range defaultClusterLocalServices {
+		defaultClusterLocalHosts = append(defaultClusterLocalHosts, host.Name(s+"."+domainSuffix))
+	}
+
+	if discoveryHost, _, err := e.GetDiscoveryAddress(); err != nil {
+		log.Errorf("failed to make discoveryAddress cluster-local: %v", err)
+	} else {
+		if !strings.HasSuffix(string(discoveryHost), domainSuffix) {
+			discoveryHost += host.Name("." + domainSuffix)
+		}
+		defaultClusterLocalHosts = append(defaultClusterLocalHosts, discoveryHost)
+	}
+
+	// Collect the cluster-local hosts.
+	clusterLocalHosts := make([]host.Name, 0)
+	for _, serviceSettings := range e.Mesh().ServiceSettings {
+		if serviceSettings.Settings.ClusterLocal {
+			for _, h := range serviceSettings.Hosts {
+				clusterLocalHosts = append(clusterLocalHosts, host.Name(h))
+			}
+		} else {
+			// Remove defaults if specified to be non-cluster-local.
+			for _, h := range serviceSettings.Hosts {
+				for i, defaultClusterLocalHost := range defaultClusterLocalHosts {
+					if len(defaultClusterLocalHost) > 0 {
+						if h == string(defaultClusterLocalHost) ||
+							(defaultClusterLocalHost.IsWildCarded() &&
+								strings.HasSuffix(h, string(defaultClusterLocalHost[1:]))) {
+							// This default was explicitly overridden, so remove it.
+							defaultClusterLocalHosts[i] = ""
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// Add any remaining defaults to the end of the list.
+	for _, defaultClusterLocalHost := range defaultClusterLocalHosts {
+		if len(defaultClusterLocalHost) > 0 {
+			clusterLocalHosts = append(clusterLocalHosts, defaultClusterLocalHost)
+		}
+	}
+
+	sort.Sort(host.Names(clusterLocalHosts))
+
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.clusterLocalHosts = clusterLocalHosts
+	return clusterLocalHosts
+}
+
+func (e *Environment) GetClusterLocalHosts() []host.Name {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	return e.clusterLocalHosts
+}
+
+func (e *Environment) IsServiceClusterLocal(service *v1.Service) bool {
+	hostname := fmt.Sprintf("%s.%s.svc.cluster.local", service.Name, service.Namespace)
```

Comment:
I think we need to lock everywhere we access `e.clusterLocalHosts`.

Reply:
nit: could you just unlock before the return rather than using `defer`? There is a slight performance penalty for using defer.